enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/slotted/SlottedPipeBuilder.scala

```diff
@@ -351,7 +352,79 @@ class SlottedPipeBuilder(fallback: PipeBuilder,
+        val rhsOverlap = overlaps(pipeline, rhsInfo)
+        if(rhsOverlap == true)
+          FastUnionSlottedPipe(lhs, rhs)(id = id)
+        else
+          UnionSlottedPipe(lhs, rhs, rhsInfo, pipeline, computeUnionSlots(lhsInfo, rhsInfo, pipeline))(id = id)
+
       case _ => throw new CantCompileQueryException(s"Unsupported logical plan operator: $plan")
     }
   }
+
+  private def overlaps(source: PipelineInformation, toCheck: PipelineInformation):Boolean = {
+    source.mapSlot {
+      case (k, s) => s == toCheck.get(k).get
+    }.forall(_ == true)
+  }
+}
+
+object SlottedPipeBuilder {
+  private def createProjectionsForResult(columns: Seq[String], pipelineInformation1: PipelineInformation): Seq[(String, Expression)] = {
+    val runtimeColumns: Seq[(String, commandExpressions.Expression)] = columns map {
+      k =>
+        pipelineInformation1(k) match {
+          case LongSlot(offset, false, CTNode, _) =>
+            k -> slottedExpressions.NodeFromSlot(offset)
+          case LongSlot(offset, true, CTNode, _) =>
+            k -> slottedExpressions.NullCheck(offset, slottedExpressions.NodeFromSlot(offset))
+          case LongSlot(offset, false, CTRelationship, _) =>
+            k -> slottedExpressions.RelationshipFromSlot(offset)
+          case LongSlot(offset, true, CTRelationship, _) =>
+            k -> slottedExpressions.NullCheck(offset, slottedExpressions.RelationshipFromSlot(offset))
+
+          case RefSlot(offset, _, _, _) =>
+            k -> slottedExpressions.ReferenceFromSlot(offset)
+
+          case _ =>
+            throw new InternalException(s"Did not find `$k` in the pipeline information")
+        }
+    }
+    runtimeColumns
+  }
+
+  //compute mapping from incoming to outgoing pipe line, the slot order may differ
+  //between the output and the input (lhs and rhs) and it may be the case that
+  //we have a reference slot in the output but a long slot on one of the inputs,
+  //e.g. MATCH (n) RETURN n UNION RETURN 42 AS n
+  def computeUnionSlots(lhsInfo: PipelineInformation, rhsInfo: PipelineInformation,
+                        out: PipelineInformation) = {
+    //find columns where output is a reference slot but where the input is a long slot
+    def findColums(input: PipelineInformation) = {
+      val slots = out.mapSlot {
+        case (k, slot: RefSlot) => input.get(k).get match {
+          case ls: LongSlot => Some(ls)
+          case _ => None
+        }
+        case _ => None
+      }.flatten.toIndexedSeq
+      slots.zip(createProjectionsForResult(slots.map(_.name), input).map(_._2)).toMap
+    }
+    //Compute how to convert input to output pipeline
+    val expressions = findColums(lhsInfo) ++ findColums(rhsInfo)
+    val mapSlots: Iterable[(ExecutionContext, ExecutionContext, PipelineInformation, QueryState) => Unit] = out.mapSlot {
+      case (k, v: LongSlot) => (in, out, info, _) =>
```

Comment:
In what way is the `info` passed in to these functions different from the `rhsInfo` we already have in scope?

Looking at the code it certainly looks like it would be the same. We pass in the same `rhsInfo` to both the `computeUnionSlots` function as we store in the `UnionSlottedPipe` object, and the `rhsInfo` stored in `UnionSlottedPipe` is passed unmodified to the functions that are called `f` there, where it ends up as `info` here.

If we instead just use `rhsInfo` here, we don't have to have `rhsInfo` available in `UnionSlottedPipe`. It also means we can move more of the decision making that happens here to compiletime. Essentially we don't need the `PipelineInformation` object referenced at all at runtime, since we can do the lookups when `computeUnionSlots` runs, and have the mappers it returns only know the actual slot offsets it is mapping (and not how to look up those slots - because that has already been done.